src/android_sdk/java/main/dev/perfetto/sdk/PerfettoTrace.java

```diff
@@ -95,37 +93,57 @@ public Category(String name, List<String> tags) {
     @CriticalNative
     private static native boolean native_is_enabled(long ptr);
 
-    @CriticalNative
-    private static native long native_get_extra_ptr(long ptr);
-
-    /** Register the category. */
-    public Category register() {
-      native_register(mPtr);
-      mIsRegistered = true;
+    /** Create the native category object and register it. */
+    public synchronized Category register() {
+      if (mPtr == 0) {
+        long ptr = native_init(mName, mTags.toArray(new String[0]));
+        sNativeMemoryCleaner.registerNativeAllocation(this, ptr, native_delete());
+        native_register(ptr);
+        // There is not much sense in the created, but not yet registered category,
+        // so we make the `ptr` visible to other threads only after registration.
+        mPtr = ptr;
+        mIsRegistered = true;
+      } else {
+        if (!mIsRegistered) {
+          native_register(mPtr);
+          mIsRegistered = true;
+        }
+      }
       return this;
     }
 
     /** Unregister the category. */
-    public Category unregister() {
-      native_unregister(mPtr);
-      mIsRegistered = false;
+    public synchronized Category unregister() {
+      if (mIsRegistered) {
+        // mIsRegistered == true implies mPtr != 0
+        mIsRegistered = false;
+        native_unregister(mPtr);
+      }
       return this;
     }
 
-    /** Whether the category is enabled or not. */
-    public boolean isEnabled() {
-      return native_is_enabled(mPtr);
+    /** Whether the category is registered and enabled or not. */
+    public boolean isRegisteredAndEnabled() {
+      // This method is explicitly not synchronized, because we want it to be as fast as possible.
+      // mIsRegistered is volatile and is set only from `#register()` or `#unregister()` methods,
+      // both of them are synchronized.
+      if (mIsRegistered) {
+        // mIsRegistered == true implies mPtr != 0
```

Comment:
I'm not sure this comment makes sense. It's very much possible for another thread to have called unregister between the previous check. I think we should make sure that we can guard against something which is fundamentally racy.

Reply:
> It's very much possible for another thread to have called unregister between the previous check.

It is fine. The `mIsRegistered` check is just a fast-path optimization to not call the native method if we know in java that the category is not registered.

but if the category is registered (or is unregistered in the process) we go to native and check in native_is_enabled the register and if it is enabled (if we have a consumer) state

Reply:
Sure please add that as a comment then to make it clear we've thought about this case. As it stands, the current comment makes it feel otherwise.

Reply:
We discussed with @zezeozue and decided to remove the `mIsRegister` check from here and move it to the `PerfettoTrackEventBuilder`, see the https://github.com/google/perfetto/pull/1481/commits/bbe0cc1d5939d169f7efe941c8bd306eab21a65e.